### 2. dev_notebook_biciPARKS stations

This is your development environment...you may start:
- Exploring your data.
- Testing the ad-hoc methods.
- Designing your pipeline.

El análisis de este notebook será más resumido, debido a las observaciones rigurosas del primer notebook **(1. dev_notebook_biciMAD stations)** y que son replicados a continuación:

In [1]:
import pandas as pd

In [2]:
biciparks_df = pd.read_csv('../data/raw/bicipark_stations.csv', sep=';') # el delimitador es ';' si abrimos el csv
biciparks_df

Unnamed: 0  stationId                 stationName  \
0            0         74  Bicipark Fuente de la Mora   
1            1         22             Bicipark Orense   
2            2         49          Bicipark Montalban   
3            3         12            Bicipark Almagro   
4            4          9          Bicipark Recoletos   
5            5          5           Bicipark Recuerdo   
6            6         25          Bicipark Salamanca   
7            7         67     Bicipark Villa de Paris   
8            8       1154              Bicipark Pitis   
9            9       1169           Bicipark Aviación   
10          10       1238          Bicipark Canalejas   

                             address     city zip_code  total_places  \
0      Estación de Fuente de la Mora   Madrid    28050            10   
1          Avenida General Perón, 27   Madrid    28020            15   
2              Calle de Montalbán, 5   Madrid    28014            16   
3                  Calle Almagro, 11   Madrid    28010            16   
4              Paseo de Recoletos, 2   Madrid    28001            10   
5                   Calle Hiedra, 26   Madrid    28036            13   
6       Pl. del Marqués de Salamanca   Madrid    28006             5   
7          Calle de Orellana, 23, 21   Madrid    28004            10   
8   Calle Pau Arroyo del Fresno, 181   Madrid    28049            10   
9          Parking Aviación Española   Madrid    28024            10   
10                                 -   Madrid        -            10   

    free_places  enabled  reserved_places geometry.type  \
0             9        1                0         Point   
1             9        1                1         Point   
2            15        1                0         Point   
3            12        1                0         Point   
4             3        0                0         Point   
5            13        1                0         Point   
6            -1        1                0         Point   
7             8        1                1         Point   
8             7        1                0         Point   
9            10        1                0         Point   
10           10        1                0         Point   

           geometry.coordinates  
0      [-3.6630679, 40.4846838]  
1        [-3.693897, 40.452973]  
2        [-3.690786, 40.418146]  
3        [-3.692879, 40.430361]  
4        [-3.692245, 40.420354]  
5        [-3.678921, 40.472419]  
6        [-3.679172, 40.430315]  
7        [-3.693698, 40.425608]  
8      [-3.7249142, 40.4959425]  
9   [-3.78369522, 40.383491516]  
10   [-3.6994506, 40.417807676]

De un vistazo, lo que me llamó la atención ha sido la estación **Bicipark Canalejas,** porque no veo información del código postal, **ni** su dirección.

A través de las coordenadas **[40.417807676, -3.6994506]** en Google Maps, encuentro "Parking Centro Canalejas, Cra de S. Jerónimo, 7, 28014 Madrid" **(para coches).** NO es un Bicipark.

He consultado la página de [EMT Madrid](https://www.emtmadrid.es/Aparcamientos/Bicipark/Disponibles?lang=es-ES) y NO existe el **Bicipark Canalejas.** Por lo tanto, me puedo cargar el registro:

In [3]:
biciparks_df = biciparks_df.drop([10])
biciparks_df.tail()

Unnamed: 0  stationId              stationName  \
5           5          5        Bicipark Recuerdo   
6           6         25       Bicipark Salamanca   
7           7         67  Bicipark Villa de Paris   
8           8       1154           Bicipark Pitis   
9           9       1169        Bicipark Aviación   

                            address    city zip_code  total_places  \
5                  Calle Hiedra, 26  Madrid    28036            13   
6      Pl. del Marqués de Salamanca  Madrid    28006             5   
7         Calle de Orellana, 23, 21  Madrid    28004            10   
8  Calle Pau Arroyo del Fresno, 181  Madrid    28049            10   
9         Parking Aviación Española  Madrid    28024            10   

   free_places  enabled  reserved_places geometry.type  \
5           13        1                0         Point   
6           -1        1                0         Point   
7            8        1                1         Point   
8            7        1                0         Point   
9           10        1                0         Point   

          geometry.coordinates  
5       [-3.678921, 40.472419]  
6       [-3.679172, 40.430315]  
7       [-3.693698, 40.425608]  
8     [-3.7249142, 40.4959425]  
9  [-3.78369522, 40.383491516]

In [4]:
biciparks_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   Unnamed: 0            10 non-null     int64 
 1   stationId             10 non-null     int64 
 2   stationName           10 non-null     object
 3   address               10 non-null     object
 4   city                  10 non-null     object
 5   zip_code              10 non-null     object
 6   total_places          10 non-null     int64 
 7   free_places           10 non-null     int64 
 8   enabled               10 non-null     int64 
 9   reserved_places       10 non-null     int64 
 10  geometry.type         10 non-null     object
 11  geometry.coordinates  10 non-null     object
dtypes: int64(6), object(6)
memory usage: 1.1+ KB


1. Tenemos **12 columnas.** Solo necesitamos **4** de ellas:

    - **stationName:** Nombre de la estación de biciPARK
    - **address:** Dirección de la estación de biciPARK
    - **zip_code:** Código postal
    - **geometry.coordinates:** Coordenadas (longitud y latitud) de la ubicación de biciPARK

Igual que en el Dataframe BiciMAD, voy a revisar que tipo de dato son algunos elementos de **geometry.coordinates**:

In [5]:
print(biciparks_df['geometry.coordinates'][0])
print(type(biciparks_df['geometry.coordinates'][0]))
print(type(biciparks_df['geometry.coordinates'][5]))

[-3.6630679, 40.4846838]
<class 'str'>
<class 'str'>


### Modificando el dataframe con las 4 columnas relevantes

In [6]:
biciparks_df = biciparks_df[['stationName','address','zip_code','geometry.coordinates']]

In [7]:
biciparks_df

stationName                           address zip_code  \
0  Bicipark Fuente de la Mora     Estación de Fuente de la Mora    28050   
1             Bicipark Orense         Avenida General Perón, 27    28020   
2          Bicipark Montalban             Calle de Montalbán, 5    28014   
3            Bicipark Almagro                 Calle Almagro, 11    28010   
4          Bicipark Recoletos             Paseo de Recoletos, 2    28001   
5           Bicipark Recuerdo                  Calle Hiedra, 26    28036   
6          Bicipark Salamanca      Pl. del Marqués de Salamanca    28006   
7     Bicipark Villa de Paris         Calle de Orellana, 23, 21    28004   
8              Bicipark Pitis  Calle Pau Arroyo del Fresno, 181    28049   
9           Bicipark Aviación         Parking Aviación Española    28024   

          geometry.coordinates  
0     [-3.6630679, 40.4846838]  
1       [-3.693897, 40.452973]  
2       [-3.690786, 40.418146]  
3       [-3.692879, 40.430361]  
4       [-3.692245, 40.420354]  
5       [-3.678921, 40.472419]  
6       [-3.679172, 40.430315]  
7       [-3.693698, 40.425608]  
8     [-3.7249142, 40.4959425]  
9  [-3.78369522, 40.383491516]

Haré el mismo tratamiento de biciMAD en la columna **geometry.coordinates:**

In [8]:
coordinates_df = biciparks_df['geometry.coordinates'].str.strip('[]').str.split(',', expand=True).astype('float64')
coordinates_df

0          1
0 -3.663068  40.484684
1 -3.693897  40.452973
2 -3.690786  40.418146
3 -3.692879  40.430361
4 -3.692245  40.420354
5 -3.678921  40.472419
6 -3.679172  40.430315
7 -3.693698  40.425608
8 -3.724914  40.495942
9 -3.783695  40.383492

In [9]:
coordinates_df.columns = ['park_long', 'park_lat']
coordinates_df

park_long   park_lat
0  -3.663068  40.484684
1  -3.693897  40.452973
2  -3.690786  40.418146
3  -3.692879  40.430361
4  -3.692245  40.420354
5  -3.678921  40.472419
6  -3.679172  40.430315
7  -3.693698  40.425608
8  -3.724914  40.495942
9  -3.783695  40.383492

**Concatenar los dataframes por columnas:**

In [10]:
biciparks_df = pd.concat([biciparks_df,coordinates_df],axis=1)
biciparks_df

stationName                           address zip_code  \
0  Bicipark Fuente de la Mora     Estación de Fuente de la Mora    28050   
1             Bicipark Orense         Avenida General Perón, 27    28020   
2          Bicipark Montalban             Calle de Montalbán, 5    28014   
3            Bicipark Almagro                 Calle Almagro, 11    28010   
4          Bicipark Recoletos             Paseo de Recoletos, 2    28001   
5           Bicipark Recuerdo                  Calle Hiedra, 26    28036   
6          Bicipark Salamanca      Pl. del Marqués de Salamanca    28006   
7     Bicipark Villa de Paris         Calle de Orellana, 23, 21    28004   
8              Bicipark Pitis  Calle Pau Arroyo del Fresno, 181    28049   
9           Bicipark Aviación         Parking Aviación Española    28024   

          geometry.coordinates  park_long   park_lat  
0     [-3.6630679, 40.4846838]  -3.663068  40.484684  
1       [-3.693897, 40.452973]  -3.693897  40.452973  
2       [-3.690786, 40.418146]  -3.690786  40.418146  
3       [-3.692879, 40.430361]  -3.692879  40.430361  
4       [-3.692245, 40.420354]  -3.692245  40.420354  
5       [-3.678921, 40.472419]  -3.678921  40.472419  
6       [-3.679172, 40.430315]  -3.679172  40.430315  
7       [-3.693698, 40.425608]  -3.693698  40.425608  
8     [-3.7249142, 40.4959425]  -3.724914  40.495942  
9  [-3.78369522, 40.383491516]  -3.783695  40.383492

In [11]:
# Elimino la columna geometry.coordinates
biciparks_df = biciparks_df.drop(columns=['geometry.coordinates'])

### Dataframe final de biciparks_df

In [12]:
# Renombramos las columnas
biciparks_df = biciparks_df.rename(columns={"stationName": "BiciPARK station",
                          "address" : "BiciPARK address"})
biciparks_df

BiciPARK station                  BiciPARK address zip_code  \
0  Bicipark Fuente de la Mora     Estación de Fuente de la Mora    28050   
1             Bicipark Orense         Avenida General Perón, 27    28020   
2          Bicipark Montalban             Calle de Montalbán, 5    28014   
3            Bicipark Almagro                 Calle Almagro, 11    28010   
4          Bicipark Recoletos             Paseo de Recoletos, 2    28001   
5           Bicipark Recuerdo                  Calle Hiedra, 26    28036   
6          Bicipark Salamanca      Pl. del Marqués de Salamanca    28006   
7     Bicipark Villa de Paris         Calle de Orellana, 23, 21    28004   
8              Bicipark Pitis  Calle Pau Arroyo del Fresno, 181    28049   
9           Bicipark Aviación         Parking Aviación Española    28024   

   park_long   park_lat  
0  -3.663068  40.484684  
1  -3.693897  40.452973  
2  -3.690786  40.418146  
3  -3.692879  40.430361  
4  -3.692245  40.420354  
5  -3.678921  40.472419  
6  -3.679172  40.430315  
7  -3.693698  40.425608  
8  -3.724914  40.495942  
9  -3.783695  40.383492

### Exportamos el .csv final de biciparks_df_v2

In [13]:
from pathlib import Path
filepath = Path('../data/processed/biciparks_df.csv')  
filepath.parent.mkdir(parents=True, exist_ok=True) 
biciparks_df.to_csv(filepath)